flink-core/src/main/java/org/apache/flink/util/AbstractCloseableRegistry.java

```diff
@@ -43,145 +39,15 @@
-     *     call.
-     */
-    public final boolean unregisterCloseable(C closeable) {
-
-        if (null == closeable) {
-            return false;
-        }
-
-        synchronized (getSynchronizationLock()) {
-            return doUnRegister(closeable, closeableToRef);
-        }
+        super(closeableToRef, true);
     }
 
     @Override
     public void close() throws IOException {
-        Collection<Closeable> toCloseCopy;
-
-        synchronized (getSynchronizationLock()) {
-            if (closed) {
-                return;
-            }
-
-            closed = true;
-
-            toCloseCopy = getReferencesToClose();
-
-            closeableToRef.clear();
-        }
-
-        IOUtils.closeAllQuietly(toCloseCopy);
-    }
-
-    public boolean isClosed() {
-        synchronized (getSynchronizationLock()) {
-            return closed;
-        }
-    }
-
-    protected Collection<Closeable> getReferencesToClose() {
-        return new ArrayList<>(closeableToRef.keySet());
-    }
-
-    /**
-     * Does the actual registration of the closeable with the registry map. This should not do any
-     * long running or potentially blocking operations as is is executed under the registry's lock.
-     */
-    protected abstract void doRegister(
-            @Nonnull C closeable, @Nonnull Map<Closeable, T> closeableMap);
-
-    /**
-     * Does the actual un-registration of the closeable from the registry map. This should not do
-     * any long running or potentially blocking operations as is is executed under the registry's
-     * lock.
-     */
-    protected abstract boolean doUnRegister(
-            @Nonnull C closeable, @Nonnull Map<Closeable, T> closeableMap);
-
-    /**
-     * Returns the lock on which manipulations to members closeableToRef and closeable must be
-     * synchronized.
-     */
-    protected final Object getSynchronizationLock() {
-        return lock;
-    }
-
-    /** Adds a mapping to the registry map, respecting locking. */
-    protected final void addCloseableInternal(Closeable closeable, T metaData) {
-        synchronized (getSynchronizationLock()) {
-            closeableToRef.put(closeable, metaData);
-        }
-    }
-
-    /** Removes a mapping from the registry map, respecting locking. */
-    protected final boolean removeCloseableInternal(Closeable closeable) {
-        synchronized (getSynchronizationLock()) {
-            return closeableToRef.remove(closeable) != null;
-        }
-    }
-
-    @VisibleForTesting
-    public final int getNumberOfRegisteredCloseables() {
-        synchronized (getSynchronizationLock()) {
-            return closeableToRef.size();
-        }
-    }
-
-    @VisibleForTesting
-    public final boolean isCloseableRegistered(Closeable c) {
-        synchronized (getSynchronizationLock()) {
-            return closeableToRef.containsKey(c);
-        }
+        IOUtils.closeAllCloseable(prepareToExceptionalClose());
```

Comment:
Something's really wrong with the entire hierarchy imo. Why do we need two abstract classes?

If we do what's the point of this close method? `prepareToExceptionalClose()` will always return an empty collection for all subclasses of this class (you pass `true` for the close quietly flag. Lastly why does the `prepareToExceptionalClose` closes anything? That's really counterintuitive.

Can't we have an abstract `doClose(Collection<R> toClose) throws E` instead?

Reply:
> Why do we need two abstract classes?

Because one of them implement `AutoCloseable` while another implements `Closeable`.

> prepareToExceptionalClose() will always return an empty collection for all subclasses of this class (you pass true for the close quietly flag. Lastly why does the prepareToExceptionalClose closes anything? 

I pass the `true` in order to keep backward compatibility with old behavior but it will be possible to change it if we need other implementation with unquiet close.
The main target of `prepareToExceptionalClose` is to avoid duplication of the code related to closing resources quietly. We can always return the list for the closing but in this case, we will need to check the `quiet` flag in both implementations. Also, we can rename it but I need to think of a good name.

> Can't we have an abstract doClose(Collection<R> toClose) throws E instead?

I wish to have it but it is not possible since `AbstractAutoCloseableRegistry` should implement `AutoCloseable` while `AbstractCloseableRegistry` should implement `Closeable`. So we can not avoid implementing `close() throws IOException`


In a conclusion, the main problem here is the implementation of two interfaces `AutoCloseable` and `Closeable` which have different exceptions for `close` and it is impossible to generalize. (maybe I am wrong here and somehow we can use generic here but I didn't find an easy way to do so)

Reply:
Sorry, but I think I am not buying into your argumentation, please take a look here: https://github.com/dawidwys/flink/commit/92bc4b1df69826bc80026c54136578a0312f1bab (I have not fixed the tests, but it should be easy to do so).

Reply:
If we want to make one implementation with quiet close while another with an exception then it will work but if we want to keep it configurable then it leads to code duplication in `close` method but maybe it is less evil. I will take a look at it close.

Reply:
If you want to still have the `AbstractCloseableRegistry` fine, I am not strong on removing this class. 

However, all other points are much simpler imo.